File: Array/zhangjunwei/linked_list/isPalindrome_6_12.java
package com.lanfairy.md;

import java.util.ArrayList;
import java.util.ListIterator;
import java.util.PriorityQueue;

public class isPalindrome {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public boolean isPalindrome1(ListNode head) {

        int len = 0;
        ListNode current = head;
        while (current != null) {
            len++;
            current = current.next;
        }
        int[] invertedequence = new int[len];
        current = head;
        int i = len - 1;
        while (current != null) {
            invertedequence[i--] = current.val;
            current = current.next;
        }

        current = head;
        i = 0;
        boolean isPalindrome = true;
        while (current != null) {
            if (current.val != invertedequence[i])
                return false;
            current = current.next;
            i++;
        }
        return true;
    }

    /****************以下是参考题解的解法*********************/
    public boolean isPalindrome(ListNode head) {
        if (head == null) return true;//[] 空链表也是回文链表
        //链表的后半部分
        ListNode firstHalf = endOfFirstHalf(head);
        //反转后半部分
       ListNode secondHalf = reverseList(firstHalf.next);
        boolean isPalindrome = true;
        ListNode fir = head;
        ListNode sec = secondHalf;
        while (sec != null) {
            if (sec.val != fir.val){
                isPalindrome = false;
                break;
            }
            sec = sec.next;
            fir = fir.next;
        }
        firstHalf.next = reverseList(secondHalf);
        return isPalindrome;
    }

    //反转链表
    private ListNode reverseList(ListNode head) {
        ListNode pre = null;
        ListNode current = head;
        while (current != null) {
            ListNode temp = current.next;
            current.next = pre;
            pre = current;
            current = temp;
        }
        return pre;
    }

    //找出后半部分
    private ListNode endOfFirstHalf(ListNode head) {
        ListNode fast = head;  //
        ListNode slow = head;
        while (fast.next != null && fast.next.next != null) {
            fast = fast.next.next;//每次走两步
            slow = slow.next;//每次走一步
        }
        return slow;
    }


}

File: Array/zhangjunwei/linked_list/isPalindrome_6_12.c
//
// Created by 张俊伟 on 2020/6/12.
//
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#define Data int
typedef struct ListNode_ {
    Data val;
    struct ListNode_ *next;
} ListNode;

bool isPalindrome1(ListNode *head) {
    int len = 0;
    ListNode *current = head;
    while (current != NULL) {
        len++;
        current = current->next;
    }
    int *rever_arr = (Data *) malloc(sizeof(Data) * len);
    current = head;
    int i = len - 1;
    while (current != NULL) {
        rever_arr[i--] = current->val;
        current = current->next;
    }
    current = head;
    i = 0;
    bool is_palindrome = true;
    while (current != NULL) {
        if (current->val != rever_arr[i]) {
            is_palindrome = false;
            break;
        }
        i++;
        current = current->next;
    }
    return is_palindrome;
}


//反转链表
ListNode *reverseList(ListNode *head) {
    ListNode *cur = head;
    ListNode *pre = NULL;
    while (cur!=NULL){
        ListNode *tmp = cur->next;
        cur->next = pre;
        pre = cur;
        cur = tmp;
    }
    return pre;
}

//找出后半部分
ListNode *endOfFirstHalf(ListNode *head) {
    ListNode *fast = head;
    ListNode *slow = head;
    while (fast->next != NULL && fast->next->next != NULL) {
        fast = fast->next->next;
        slow = slow->next;
    }
    return slow;
}
bool isPalindrome(ListNode *head) {
    if(head==NULL) return true;//[] 空链表也是回文链表
    ListNode *fir_half = endOfFirstHalf(head);
    ListNode *sec_half = reverseList(fir_half->next);
    ListNode *p1 = head;
    ListNode *p2 = sec_half;
    bool is_pal = true;
    while (p2!=NULL){
        if (p1->val!=p2->val)
            is_pal = false;
        p1 = p1->next;
        p2 = p2->next;
    }
    //将反转的后半部分链表 再次反转复原
    fir_half->next = reverseList(sec_half);
    return is_pal;
}
